### PR TITLE
fix: android download button

### PR DESCRIPTION
**Pull request by zhouhanseng**
close https://github.com/DimensionDev/Maskbook/issues/1952

### CODE REVIEW

File: packages/maskbook/src/extension/options-page/DashboardDialogs/Backup.tsx [hunk 1]
```diff
@@ -36,9 +36,13 @@ const useDatabaseStyles = makeStyles((theme: Theme) =>
 export function DashboardBackupDialog(props: WrappedDialogProps) {
     const { t } = useI18N()
     const classes = useDatabaseStyles()
-    const { enqueueSnackbar, closeSnackbar } = useSnackbar()
+    const { enqueueSnackbar } = useSnackbar()
 
     const { value, loading } = useAsync(() => Services.Welcome.generateBackupJSON())
+
+    // since Android doesn't support `browser.download.downloads`,
+    //  we should create download url before click button.
```
Comment (zhouhanseng) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/Backup.tsx:44]: Like how we get the download image url in `<AutoPasteFailedDialog />`

File: packages/maskbook/src/extension/background-script/HelperService.ts [hunk 2]
```diff
@@ -27,21 +25,12 @@ export function fetch(url: string) {
     return globalThis.fetch(url).then((x) => x.blob())
 }
 
-const createDlElement = once(() => document.createElement('a'))
```
Comment (zhouhanseng) [inline packages/maskbook/src/extension/background-script/HelperService.ts:30]: Removed, this way doesn't work on Android.

File: packages/maskbook/src/extension/options-page/DashboardDialogs/Backup.tsx [hunk 3]
```diff
@@ -36,9 +36,13 @@ const useDatabaseStyles = makeStyles((theme: Theme) =>
 export function DashboardBackupDialog(props: WrappedDialogProps) {
     const { t } = useI18N()
     const classes = useDatabaseStyles()
-    const { enqueueSnackbar, closeSnackbar } = useSnackbar()
```
Comment (zhouhanseng) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/Backup.tsx:39]: unused code

File: packages/maskbook/src/extension/options-page/DashboardDialogs/Backup.tsx [hunk 4]
```diff
@@ -36,9 +36,13 @@ const useDatabaseStyles = makeStyles((theme: Theme) =>
...
+    const { enqueueSnackbar } = useSnackbar()
 
     const { value, loading } = useAsync(() => Services.Welcome.generateBackupJSON())
+
+    // since Android doesn't support `browser.download.downloads`,
+    //  we should create download url before click button.
+    const backupInfo = useAsync(() => Services.Welcome.createBackupUrl({ download: false, onlyBackupWhoAmI: false }))
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/Backup.tsx:45]: Isn't it costly?

Comment (zhouhanseng) [reply packages/maskbook/src/extension/options-page/DashboardDialogs/Backup.tsx:45]: That's right, it cost. But what else can we do?

**Comment by jk234ert** [comment]
Still not work on Android

**Review by Jack-Works — APPROVED** [review]

**Comment by zhouhanseng** [comment]
@Jack-Works Revert to the costly and un-dry but worked way temporarily since `document.createElement('a')` does not work on Android webview (IDK why). In the next release of Mobile app, we will get rid of this by using new message channel.

File: packages/maskbook/src/components/InjectedComponents/AutoPasteFailedDialog.tsx [hunk 5]
```diff
@@ -116,12 +116,23 @@ export function AutoPasteFailedDialog(props: AutoPasteFailedDialogProps) {
                             </Button>
                         ) : null}
                         {url ? (
-                            <Button
-                                variant="text"
-                                onClick={() => saveAsFileFromUrl(url, fileName)}
-                                startIcon={<Download />}>
-                                {t('download')}
-                            </Button>
+                            process.env.architecture === 'app' && process.env.target === 'firefox' ? (
```
Comment (septs) [inline packages/maskbook/src/components/InjectedComponents/AutoPasteFailedDialog.tsx:119]: extract to a common component, it is reusable.

e.q: `packages/maskbook/src/utils/components/DownloadButton.tsx`

CC @Jack-Works

Comment (zhouhanseng) [reply packages/maskbook/src/components/InjectedComponents/AutoPasteFailedDialog.tsx:119]: Hi, this is temporary workaround, this part will be simplified later.

Comment (septs) [reply packages/maskbook/src/components/InjectedComponents/AutoPasteFailedDialog.tsx:119]: add an `// HACK` comments, thanks.

Comment (zhouhanseng) [reply packages/maskbook/src/components/InjectedComponents/AutoPasteFailedDialog.tsx:119]: ok

Comment (Jack-Works) [reply packages/maskbook/src/components/InjectedComponents/AutoPasteFailedDialog.tsx:119]: ACK

File: packages/maskbook/src/extension/options-page/DashboardDialogs/Backup.tsx (hunk 4 above)
Comment (Jack-Works) [reply packages/maskbook/src/extension/options-page/DashboardDialogs/Backup.tsx:48]: this costy op only used in android but it will be run on any platform

**Review by jk234ert — APPROVED** [review]

**Review by Jack-Works — APPROVED** [review]

**Pull request merged**